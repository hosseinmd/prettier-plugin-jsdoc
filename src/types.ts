import { ParserOptions } from "prettier";

export interface JsdocOptions {
  jsdocSpaces: number;
  jsdocPrintWidth?: number;
  jsdocDescriptionWithDot: boolean;
  jsdocDescriptionTag: boolean;
  jsdocVerticalAlignment: boolean;
  jsdocKeepUnParseAbleExampleIndent: boolean;
  /** default is true */
  jsdocSingleLineComment: boolean;
  jsdocSeparateReturnsFromParam: boolean;
  jsdocSeparateTagGroups: boolean;
  jsdocAddDefaultToDescription: boolean;
  jsdocCapitalizeDescription: boolean;
  tsdoc: boolean;
}

export interface AllOptions extends ParserOptions, JsdocOptions {}

type LocationDetails = { line: number; column: number };
type Location = { start: LocationDetails; end: LocationDetails };

export type PrettierComment = {
  type: "CommentBlock" | "Block";
  value: string;
  start: number;
  end: number;
  loc: Location;
};

export type Token = {
  type:
    | "CommentBlock"
    | "Block"
    | {
        label: string; // "function" | "name";
        keyword?: string;
        beforeExpr: boolean;
        startsExpr: boolean;
        rightAssociative: boolean;
        isLoop: boolean;
        isAssign: boolean;
        prefix: boolean;
        postfix: boolean;
        binop: null;
      };
  value: string;
  start: number;
  end: number;
  loc: Location;
};

export type AST = {
  start: number;
  end: number;
  loc: Location;
  errors: [];
  program: {
    type: "Program";
    start: number;
    end: number;
    loc: [];
    sourceType: "module";
    interpreter: null;
    body: [];
    directives: [];
  };
  comments: PrettierComment[];
  tokens: Token[];
};
