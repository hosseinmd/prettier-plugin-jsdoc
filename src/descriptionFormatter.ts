import { format, BuiltInParserName } from "prettier";
import { DESCRIPTION, EXAMPLE, TODO } from "./tags";
import { AllOptions } from "./types";
import { capitalizer, formatCode } from "./utils";
import fromMarkdown from "mdast-util-from-markdown";
import { Root, Content, Link, Image, Text, List } from "mdast";

const TABLE = "2@^5!~#sdE!_TABLE";

interface DescriptionEndLineParams {
  tag: string;
  isEndTag: boolean;
}

const parserSynonyms = (lang: string): BuiltInParserName[] => {
  switch (lang) {
    case "js":
    case "javascript":
    case "jsx":
      return ["babel", "babel-flow", "vue"];
    case "ts":
    case "typescript":
    case "tsx":
      return ["typescript", "babel-ts", "angular"];
    case "json":
    case "css":
      return ["css"];
    case "less":
      return ["less"];
    case "scss":
      return ["scss"];
    case "html":
      return ["html"];
    case "yaml":
      return ["yaml"];
    default:
      return ["babel"];
  }
};

function descriptionEndLine({
  tag,
  isEndTag,
}: DescriptionEndLineParams): string {
  if ([DESCRIPTION, EXAMPLE, TODO].includes(tag) && !isEndTag) {
    return "\n";
  }

  return "";
}

interface FormatOptions {
  tagStringLength?: number;
  beginningSpace: string;
}

/**
 * Trim, make single line with capitalized text. Insert dot if flag for it is
 * set to true and last character is a word character
 *
 * @private
 */
function formatDescription(
  tag: string,
  text: string,
  options: AllOptions,
  formatOptions: FormatOptions,
): string {
  if (!text) return text;

  const { printWidth } = options;
  const { tagStringLength = 0, beginningSpace } = formatOptions;

  /**
   * change list with dash to dot for example:
   * 1- a thing
   *
   * to
   *
   * 1. a thing
   */
  text = text.replace(/^(\d+)[-][\s|]+/g, "$1. "); // Start
  text = text.replace(/\n+(\s*\d+)[-][\s]+/g, "\n$1. ");

  const fencedCodeBlocks = text.matchAll(/```\S*?\n[\s\S]+?```/gm);
  const indentedCodeBlocks = text.matchAll(
    /^\r?\n^(?:(?:(?:[ ]{4}|\t).*(?:\r?\n|$))+)/gm,
  );
  const allCodeBlocks = [...fencedCodeBlocks, ...indentedCodeBlocks];
  const tables: string[] = [];
  text = text.replace(
    /((\n|^)\|[\s\S]*?)((\n[^|])|$)/g,
    (code, _1, _2, _3, _, offs: number) => {
      // If this potential table is inside a code block, don't touch it
      for (const block of allCodeBlocks) {
        if (
          block.index !== undefined &&
          block.index <= offs + 1 &&
          offs + code.length + 1 <= block.index + block[0].length
        ) {
          return code;
        }
      }

      code = _3 ? code.slice(0, -1) : code;

      tables.push(code);
      return `\n\n${TABLE}\n\n${_3 ? _3.slice(1) : ""}`;
    },
  );
  if (options.jsdocCapitalizeDescription) text = capitalizer(text);

  text = `${tagStringLength ? `${"!".repeat(tagStringLength - 1)}?` : ""}${
    text.startsWith("```") ? "\n" : ""
  }${text}`;

  let tableIndex = 0;

  const rootAst = fromMarkdown(text);

  function stringifyASTWithoutChildren(
    mdAst: Content | Root,
    intention: string,
    parent: Content | Root | null,
  ) {
    if (mdAst.type === "inlineCode") {
      return `\`${mdAst.value}\``;
    }

    if (mdAst.type === "code") {
      let result = mdAst.value || "";
      let _intention = intention;

      if (result) {
        // Remove two space from lines, maybe added previous format
        if (mdAst.lang) {
          const supportParsers = parserSynonyms(mdAst.lang.toLowerCase());
          const parser = supportParsers?.includes(options.parser as any)
            ? options.parser
            : supportParsers?.[0] || mdAst.lang;

          result = formatCode(result, intention, {
            ...options,
            parser,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        } else if (options.jsdocPreferCodeFences || false) {
          result = formatCode(result, _intention, {
            ...options,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        } else {
          _intention = intention + " ".repeat(4);

          result = formatCode(result, _intention, {
            ...options,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        }
      }
      const addFence = options.jsdocPreferCodeFences || !!mdAst.lang;
      result = addFence ? result : result.trimEnd();
      return result
        ? addFence
          ? `\n\n${_intention}\`\`\`${mdAst.lang || ""}${result}\`\`\``
          : `\n${result}`
        : "";
    }

    if ((mdAst as Text).value === TABLE) {
      if (parent) {
        (parent as any).costumeType = TABLE;
      }

      if (tables.length > 0) {
        let result = tables?.[tableIndex] || "";
        tableIndex++;
        if (result) {
          result = format(result, {
            ...options,
            parser: "markdown",
          }).trim();
        }
        return `${
          result
            ? `\n\n${intention}${result.split("\n").join(`\n${intention}`)}`
            : (mdAst as Text).value
        }`;
      }
    }

    if (mdAst.type === "break") {
      return `\\\n`;
    }

    return ((mdAst as Text).value ||
      (mdAst as Link).title ||
      (mdAst as Image).alt ||
      "") as string;
  }

  function stringyfy(
    mdAst: Content | Root,
    intention: string,
    parent: Content | Root | null,
  ): string {
    if (!Array.isArray((mdAst as Root).children)) {
      return stringifyASTWithoutChildren(mdAst, intention, parent);
    }

    return ((mdAst as Root).children as Content[])
      .map((ast, index) => {
        if (ast.type === "listItem") {
          let _listCount = `\n${intention}- `;
          // .replace(/((?!(^))\n)/g, "\n" + _intention);
          if (typeof (mdAst as List).start === "number") {
            const count = index + (((mdAst as List).start as number) ?? 1);
            _listCount = `\n${intention}${count}. `;
          }

          const _intention = intention + " ".repeat(_listCount.length - 1);

          const result = stringyfy(ast, _intention, mdAst).trim();

          return `${_listCount}${result}`;
        }

        if (ast.type === "list") {
          let end = "";
          /**
           * Add empty line after list if that is end of description
           * issue: {@link https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/98}
           */
          if (
            tag !== DESCRIPTION &&
            mdAst.type === "root" &&
            index === mdAst.children.length - 1
          ) {
            end = "\n";
          }
          return `\n${stringyfy(ast, intention, mdAst)}${end}`;
        }

        if (ast.type === "paragraph") {
          const paragraph = stringyfy(ast, intention, parent);
          if ((ast as any).costumeType === TABLE) {
            return paragraph;
          }

          return `\n\n${paragraph
            /**
             * Break by backslash\
             * issue: https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/102
             */
            .split("\\\n")
            .map((_paragraph) => {
              const links: string[] = [];
              // Find jsdoc links and remove spaces
              _paragraph = _paragraph.replace(
                /{@link[\s](([^{}])*)}/g,
                (_, link: string) => {
                  links.push(link);

                  return `{@link${"_".repeat(link.length)}}`;
                },
              );

              _paragraph = _paragraph.replace(/\s+/g, " "); // Make single line

              if (options.jsdocCapitalizeDescription)
                _paragraph = capitalizer(_paragraph);
              if (options.jsdocDescriptionWithDot)
                _paragraph = _paragraph.replace(/([\w\p{L}])$/u, "$1."); // Insert dot if needed

              let result = breakDescriptionToLines(
                _paragraph,
                printWidth,
                intention,
              );

              // Replace links
              result = result.replace(
                /{@link([_]+)}/g,
                (original: string, underline: string) => {
                  const link = links[0];

                  if (link.length === underline.length) {
                    links.shift();
                    return `{@link ${link}}`;
                  }

                  return original;
                },
              );

              return result;
            })
            .join("\\\n")}`;
        }

        if (ast.type === "strong") {
          return `**${stringyfy(ast, intention, mdAst)}**`;
        }

        if (ast.type === "emphasis") {
          return `_${stringyfy(ast, intention, mdAst)}_`;
        }

        if (ast.type === "heading") {
          return `\n\n${intention}${"#".repeat(ast.depth)} ${stringyfy(
            ast,
            intention,
            mdAst,
          )}`;
        }

        if (ast.type === "link" || ast.type === "image") {
          return `[${stringyfy(ast, intention, mdAst)}](${ast.url})`;
        }

        return stringyfy(ast, intention, mdAst);
      })
      .join("");
  }

  let result = stringyfy(rootAst, beginningSpace, null);

  result = result.replace(/^[\s\n]+/g, "");
  result = result.replace(/^([!]+\?)/g, "");

  return result;
}

function breakDescriptionToLines(
  desContent: string,
  maxWidth: number,
  beginningSpace: string,
): string {
  let str = desContent.trim();

  if (!str) {
    return str;
  }

  const extraLastLineWidth = 10;
  let result = "";
  while (str.length > maxWidth + extraLastLineWidth) {
    let sliceIndex = str.lastIndexOf(
      " ",
      str.startsWith("\n") ? maxWidth + 1 : maxWidth,
    );
    /**
     * When a str is a long word lastIndexOf will gives 4 every time loop
     * running unlimited time
     */
    if (sliceIndex <= beginningSpace.length)
      sliceIndex = str.indexOf(" ", beginningSpace.length + 1);

    if (sliceIndex === -1) sliceIndex = str.length;

    result += str.substring(0, sliceIndex);
    str = str.substring(sliceIndex + 1);
    if (str) {
      str = `${beginningSpace}${str}`;
      str = `\n${str}`;
    }
  }

  result += str;

  return `${beginningSpace}${result}`;
}

export { descriptionEndLine, FormatOptions, formatDescription };
