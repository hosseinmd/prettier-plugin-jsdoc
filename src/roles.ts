import {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  MEMBER,
  MEMBEROF,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PROPERTY,
  PROVIDES_MODULE,
  RETURNS,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPEDEF,
  VERSION,
  YIELDS,
} from "./tags";

const TAGS_SYNONYMS = {
  // One TAG TYPE can have different titles called SYNONYMS.  We want
  // to avoid different titles in the same tag so here is map with
  // synonyms as keys and tag type as value that we want to have in
  // final jsDoc.
  arg: PARAM,
  argument: PARAM,
  const: CONSTANT,
  constructor: CLASS,
  defaultvalue: DEFAULT,
  desc: DESCRIPTION,
  emits: FIRES,
  examples: EXAMPLE,
  exception: THROWS,
  fileoverview: FILE,
  func: FUNCTION,
  host: EXTERNAL,
  method: FUNCTION,
  overview: FILE,
  params: PARAM,
  prop: PROPERTY,
  return: RETURNS,
  var: MEMBER,
  virtual: ABSTRACT,
  yield: YIELDS,
};

const TAGS_NAMELESS = [
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  RETURNS,
  THROWS,
  TODO,
  YIELDS,
  DEPRECATED,
  OVERRIDE,
  SINCE,
];

const TAGS_TYPELESS = [DESCRIPTION, EXAMPLE, TODO, DEPRECATED, OVERRIDE, SINCE];

const TAGS_DESCRIPTION_NEEDED = [CATEGORY, DESCRIPTION, EXAMPLE, SINCE, TODO];
const TAGS_NEED_FORMAT_DESCRIPTION = [
  CATEGORY,
  DESCRIPTION,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TODO,
  TYPE,
  TYPEDEF,
  YIELDS,
  DEPRECATED,
  OVERRIDE,
];
const TAGS_TYPE_NEEDED = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  TYPEDEF,
  YIELDS,
];

const TAGS_VERTICALLY_ALIGN_ABLE = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  TYPEDEF,
  YIELDS,
];

const TAGS_IS_CAMEL_CASE = [PROVIDES_MODULE];
const TAGS_GROUP = [CALLBACK, TYPEDEF];

const TAGS_ORDER = [
  PROVIDES_MODULE,
  FLOW,
  ASYNC,
  PRIVATE,
  MEMBEROF,
  VERSION,
  AUTHOR,
  DEPRECATED,
  SINCE,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  ABSTRACT,
  AUGMENTS,
  CONSTANT,
  DEFAULT,
  EXTERNAL,
  FILE,
  FIRES,
  TEMPLATE,
  FUNCTION,
  CLASS,
  EXTENDS,
  MEMBER,
  TYPEDEF,
  TYPE,
  PROPERTY,
  CALLBACK,
  PARAM,
  YIELDS,
  RETURNS,
  THROWS,
  "other",
  SEE,
  TODO,
];

export {
  TAGS_DESCRIPTION_NEEDED,
  TAGS_IS_CAMEL_CASE,
  TAGS_NAMELESS,
  TAGS_NEED_FORMAT_DESCRIPTION,
  TAGS_GROUP,
  TAGS_ORDER,
  TAGS_SYNONYMS,
  TAGS_TYPE_NEEDED,
  TAGS_TYPELESS,
  TAGS_VERTICALLY_ALIGN_ABLE,
};
