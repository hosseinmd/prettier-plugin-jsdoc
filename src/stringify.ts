import { Tag } from "comment-parser";
import { format } from "prettier";
import { formatDescription, descriptionEndLine } from "./descriptionFormatter";
import { DESCRIPTION, EXAMPLE, MEMBEROF, SEE, SPACE_TAG_DATA } from "./tags";
import { TAGS_VERTICALLY_ALIGN_ABLE } from "./roles";
import { JsdocOptions } from "./types";

const stringify = (
  { name, description, type, tag }: Tag,
  tagIndex: number,
  finalTagsArray: Tag[],
  options: JsdocOptions,
  maxTagTitleLength: number,
  maxTagTypeNameLength: number,
  maxTagNameLength: number,
): string => {
  let tagString = "\n";

  if (tag === SPACE_TAG_DATA.tag) {
    return tagString;
  }

  const {
    printWidth,
    jsdocSpaces,
    jsdocVerticalAlignment,
    jsdocDescriptionTag,
    jsdocKeepUnParseAbleExampleIndent,
  } = options;
  const gap = " ".repeat(jsdocSpaces);

  let tagTitleGapAdj = 0;
  let tagTypeGapAdj = 0;
  let tagNameGapAdj = 0;
  let descGapAdj = 0;

  if (jsdocVerticalAlignment && TAGS_VERTICALLY_ALIGN_ABLE.includes(tag)) {
    if (tag) tagTitleGapAdj += maxTagTitleLength - tag.length;
    else if (maxTagTitleLength) descGapAdj += maxTagTitleLength + gap.length;

    if (type) tagTypeGapAdj += maxTagTypeNameLength - type.length;
    else if (maxTagTypeNameLength)
      descGapAdj += maxTagTypeNameLength + gap.length;

    if (name) tagNameGapAdj += maxTagNameLength - name.length;
    else if (maxTagNameLength) descGapAdj = maxTagNameLength + gap.length;
  }

  const useTagTitle = tag !== DESCRIPTION || jsdocDescriptionTag;

  if (useTagTitle) {
    tagString += `@${tag}${" ".repeat(tagTitleGapAdj || 0)}`;
  }
  if (type) {
    tagString += gap + `{${type}}` + " ".repeat(tagTypeGapAdj);
  }
  if (name) tagString += `${gap}${name}${" ".repeat(tagNameGapAdj)}`;

  // Add description (complicated because of text wrap)
  if (description && tag !== EXAMPLE) {
    if (useTagTitle) tagString += gap + " ".repeat(descGapAdj);
    if ([MEMBEROF, SEE].includes(tag)) {
      // Avoid wrapping
      tagString += description;
    } else {
      const [, firstWord] = /^\s*(\S+)/.exec(description) || ["", ""];
      if (tagString.length + firstWord.length > printWidth) {
        // the tag is already longer than we are allowed to, so let's start at a new line
        tagString += "\n  " + formatDescription(tag, description, options);
      } else {
        // append the description to the tag
        tagString += formatDescription(tag, description, options, {
          firstLinePrintWidth: printWidth - tagString.length,
        });
      }
    }
  }

  // Try to use prettier on @example tag description
  if (tag === EXAMPLE) {
    const exampleCaption = description.match(/<caption>([\s\S]*?)<\/caption>/i);

    if (exampleCaption) {
      description = description.replace(exampleCaption[0], "");
      tagString = `${tagString} ${exampleCaption[0]}`;
    }

    try {
      let formattedExample = "";
      const examplePrintWith = printWidth - "  ".length;

      description = description.replace(/\n[^\S\r\n]{2}/g, "\n"); // Remove two space from lines, maybe added previous format

      // If example is a json
      if (description.trim().startsWith("{")) {
        formattedExample = format(description || "", {
          ...options,
          parser: "json",
          printWidth: examplePrintWith,
        });
      } else {
        formattedExample = format(description || "", {
          ...options,
          printWidth: examplePrintWith,
        });
      }

      tagString += formattedExample.replace(/(^|\n)/g, "\n  "); // Add tow space to start of lines
      tagString = tagString.slice(0, tagString.length - 3);
    } catch (err) {
      tagString += "\n";
      tagString += description
        .split("\n")
        .map((l) => `  ${jsdocKeepUnParseAbleExampleIndent ? l : l.trim()}`)
        .join("\n");
    }
  }

  // Add empty line after some tags if there is something below
  tagString += descriptionEndLine({
    description: tagString,
    tag,
    isEndTag: tagIndex === finalTagsArray.length - 1,
  });

  return tagString;
};

export { stringify };
