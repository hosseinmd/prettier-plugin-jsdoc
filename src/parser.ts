import commentParser from "comment-parser";
import {
  addStarsToTheBeginningOfTheLines,
  convertToModernType,
  formatType,
  detectEndOfLine,
} from "./utils";
import { DESCRIPTION } from "./tags";
import {
  TAGS_DESCRIPTION_NEEDED,
  TAGS_GROUP,
  TAGS_IS_CAMEL_CASE,
  TAGS_NAMELESS,
  TAGS_ORDER,
  TAGS_SYNONYMS,
  TAGS_TYPELESS,
  TAGS_VERTICALLY_ALIGN_ABLE,
} from "./roles";
import { AST, JsdocOptions, PrettierComment } from "./types";
import { stringify } from "./stringify";
import { convertCommentDescToDescTag } from "./descriptionFormatter";
import { Parser } from "prettier";
import { SPACE_TAG_DATA } from "./tags";

/** @link https://prettier.io/docs/en/api.html#custom-parser-api} */
export const getParser = (parser: Parser["parse"]) =>
  function jsdocParser(
    text: string,
    parsers: Parameters<Parser["parse"]>[1],
    options: JsdocOptions,
  ): AST {
    const ast = parser(text, parsers, options) as AST;

    if (!options.jsdocParser) {
      return ast;
    }

    const eol =
      options.endOfLine === "auto" ? detectEndOfLine(text) : options.endOfLine;
    options = { ...options, endOfLine: "lf" };

    /**
     * Control order of tags by weights. Smaller value brings tag higher.
     *
     * @param {String} tagTitle TODO
     * @returns {Number} Tag weight
     */
    function getTagOrderWeight(tag: string): number {
      if (tag === DESCRIPTION && !options.jsdocDescriptionTag) {
        return -1;
      }
      const index = TAGS_ORDER.indexOf(tag);
      return index === -1 ? TAGS_ORDER.indexOf("other") : index;
    }

    ast.comments.forEach((comment) => {
      if (!isBlockComment(comment)) return;

      /** Issue: https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/18 */
      comment.value = comment.value.replace(/^([*]+)/g, "*");

      // Create the full comment string with line ends normalized to \n
      // This means that all following code can assume \n and should only use
      // \n.
      const commentString = `/*${comment.value.replace(/\r\n?/g, "\n")}*/`;

      /**
       * Check if this comment block is a JSDoc. Based on:
       * https://github.com/jsdoc/jsdoc/blob/master/packages/jsdoc/plugins/commentsOnly.js
       */
      if (!/^\/\*\*[\s\S]+?\*\/$/.test(commentString)) return;

      const parsed = commentParser(commentString, {
        dotted_names: false,
        trim: false,
      })[0];

      if (!parsed) {
        // Error on commentParser
        return;
      }

      comment.value = "";

      convertCommentDescToDescTag(parsed);

      let maxTagTitleLength = 0;
      let maxTagTypeNameLength = 0;
      let maxTagNameLength = 0;

      parsed.tags
        // Prepare tags data
        .map(
          ({
            name,
            description,
            type,
            tag,
            source,
            optional,
            default: _default,
            ...restInfo
          }) => {
            /** When space between tag and type missed */
            const tagSticksToType = tag.indexOf("{");
            if (tagSticksToType !== -1 && tag[tag.length - 1] === "}") {
              type = tag.slice(tagSticksToType + 1, -1);
              tag = tag.slice(0, tagSticksToType);
            }

            if (TAGS_ORDER.includes(tag) && !TAGS_IS_CAMEL_CASE.includes(tag)) {
              tag = tag && tag.trim().toLowerCase();
            }

            if (tag in TAGS_SYNONYMS) {
              tag = TAGS_SYNONYMS[tag as keyof typeof TAGS_SYNONYMS];
            }

            const isVerticallyAlignAbleTags = TAGS_VERTICALLY_ALIGN_ABLE.includes(
              tag,
            );

            if (TAGS_NAMELESS.includes(tag) && name) {
              description = `${name} ${description}`;
              name = "";
            }

            if (TAGS_TYPELESS.includes(tag) && type) {
              description = `{${type}} ${description}`;
              type = "";
            }

            if (type) {
              /**
               * Convert optional to standard
               * https://jsdoc.app/tags-type.html#:~:text=Optional%20parameter
               */
              type = type.replace(/[=]$/, () => {
                optional = true;
                return "";
              });

              type = convertToModernType(type);
              type = formatType(type, options);

              if (isVerticallyAlignAbleTags)
                maxTagTypeNameLength = Math.max(
                  maxTagTypeNameLength,
                  type.length,
                );

              // Additional operations on name
              if (name) {
                // Optional tag name
                if (optional) {
                  // Figure out if tag type have default value
                  _default = _default?.trim();
                  if (_default) {
                    description = description
                      .trim()
                      .replace(/[ \t]*Default is `.*`\.?$/, "");
                    if (description && !/[.\n]$/.test(description)) {
                      description += ".";
                    }
                    description += ` Default is \`${_default}\``;
                    name = `[${name}=${_default}]`;
                  } else {
                    name = `[${name}]`;
                  }
                }

                if (isVerticallyAlignAbleTags)
                  maxTagNameLength = Math.max(maxTagNameLength, name.length);
              }
            }

            if (isVerticallyAlignAbleTags) {
              maxTagTitleLength = Math.max(maxTagTitleLength, tag.length);
            }

            description = description || "";
            description = description.trim();

            return {
              ...restInfo,
              name,
              description,
              type,
              tag,
              source,
              default: _default,
              optional,
            } as commentParser.Tag;
          },
        )

        // Group tags
        .reduce<commentParser.Tag[][]>((tagGroups, cur) => {
          if (tagGroups.length === 0 || TAGS_GROUP.includes(cur.tag)) {
            tagGroups.push([]);
          }
          tagGroups[tagGroups.length - 1].push(cur);

          return tagGroups;
        }, [])
        .flatMap((tagGroup, index, tags) => {
          // sort tags within groups
          tagGroup.sort((a, b) => {
            return getTagOrderWeight(a.tag) - getTagOrderWeight(b.tag);
          });

          // add an empty line between groups
          if (tags.length - 1 !== index) {
            tagGroup.push(SPACE_TAG_DATA);
          }

          return tagGroup;
        })
        .filter(({ description, tag }) => {
          if (!description && TAGS_DESCRIPTION_NEEDED.includes(tag)) {
            return false;
          }
          return true;
        })
        // Create final jsDoc string
        .forEach((tagData, tagIndex, finalTagsArray) => {
          comment.value += stringify(
            tagData,
            tagIndex,
            finalTagsArray,
            options,
            comment,
            maxTagTitleLength,
            maxTagTypeNameLength,
            maxTagNameLength,
          );
        });

      comment.value = comment.value.trimEnd();
      if (comment.value) {
        comment.value = addStarsToTheBeginningOfTheLines(comment.value);
      }

      if (eol === "cr") {
        comment.value = comment.value.replace(/\n/g, "\r");
      } else if (eol === "crlf") {
        comment.value = comment.value.replace(/\n/g, "\r\n");
      }
    });

    ast.comments = ast.comments.filter(
      (comment) => !(isBlockComment(comment) && !comment.value),
    );

    return ast;
  };

function isBlockComment(comment: PrettierComment): boolean {
  return comment.type === "CommentBlock" || comment.type === "Block";
}
