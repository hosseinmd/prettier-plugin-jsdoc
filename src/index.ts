import { getParser } from "./parser";
import parserBabel from "prettier/parser-babel";
import parserFlow from "prettier/parser-flow";
import parserTypescript from "prettier/parser-typescript";
import prettier, { SupportOption } from "prettier";
import { JsdocOptions } from "./types";
import { findPluginByParser } from "./utils";

const options: Record<keyof JsdocOptions, SupportOption> = {
  jsdocSpaces: {
    name: "jsdocSpaces",
    type: "int",
    category: "jsdoc",
    default: 1,
    description: "How many spaces will be used to separate tag elements.",
  },
  jsdocDescriptionWithDot: {
    name: "jsdocDescriptionWithDot",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should dot be inserted at the end of description",
  },
  jsdocDescriptionTag: {
    name: "jsdocDescriptionTag",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should description tag be used",
  },
  jsdocVerticalAlignment: {
    name: "jsdocVerticalAlignment",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should tags, types, names and description be aligned",
  },
  jsdocKeepUnParseAbleExampleIndent: {
    name: "jsdocKeepUnParseAbleExampleIndent",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description:
      "Should unParseAble example (pseudo code or no js code) keep its indentation",
  },
  jsdocSingleLineComment: {
    name: "jsdocSingleLineComment",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Should compact single line comment",
  },
  jsdocSeparateReturnsFromParam: {
    name: "jsdocSeparateReturnsFromParam",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Add an space between last @param and @returns",
  },
  jsdocCapitalizeDescription: {
    name: "jsdocCapitalizeDescription",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Should capitalize first letter of description",
  },
  tsdoc: {
    name: "tsdoc",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should format as tsdoc",
  },
  jsdocPrintWidth: {
    name: "jsdocPrintWidth",
    type: "int",
    category: "jsdoc",
    default: undefined as any,
    description:
      "If You don't set value to jsdocPrintWidth, the printWidth will be use as jsdocPrintWidth.",
  },
};

const defaultOptions: JsdocOptions = {
  jsdocSpaces: options.jsdocSpaces.default as number,
  jsdocPrintWidth: (options.jsdocPrintWidth.default as unknown) as undefined,
  jsdocDescriptionWithDot: options.jsdocDescriptionWithDot.default as boolean,
  jsdocDescriptionTag: options.jsdocDescriptionTag.default as boolean,
  jsdocVerticalAlignment: options.jsdocVerticalAlignment.default as boolean,
  jsdocKeepUnParseAbleExampleIndent: options.jsdocKeepUnParseAbleExampleIndent
    .default as boolean,
  jsdocSingleLineComment: options.jsdocSingleLineComment.default as boolean,
  jsdocSeparateReturnsFromParam: options.jsdocSeparateReturnsFromParam
    .default as boolean,
  jsdocCapitalizeDescription: options.jsdocCapitalizeDescription
    .default as boolean,
  tsdoc: options.tsdoc.default as boolean,
};

const languages = prettier
  .getSupportInfo()
  .languages.filter(({ name }) =>
    [
      "JavaScript",
      "Flow",
      "JSX",
      "TSX",
      "TypeScript",
      "Markdown",
      "MDX",
    ].includes(name),
  );

const parsers = {
  // JS - Babel
  get babel() {
    const parser = parserBabel.parsers.babel;
    return mergeParsers(parser, "babel");
  },
  get "babel-flow"() {
    const parser = parserBabel.parsers["babel-flow"];
    return mergeParsers(parser, "babel-flow");
  },
  get "babel-ts"() {
    const parser = parserBabel.parsers["babel-ts"];
    return mergeParsers(parser, "babel-ts");
  },
  // JS - Flow
  get flow() {
    const parser = parserFlow.parsers.flow;
    return mergeParsers(parser, "flow");
  },
  // JS - TypeScript
  get typescript(): prettier.Parser {
    const parser = parserTypescript.parsers.typescript;

    return mergeParsers(parser, "typescript");
    // require("./parser-typescript").parsers.typescript;
  },
  get "jsdoc-parser"() {
    // Backward compatible, don't use this in new version since 1.0.0
    const parser = parserBabel.parsers["babel-ts"];

    return mergeParsers(parser, "babel-ts");
  },
};

function mergeParsers(originalParser: prettier.Parser, parserName: string) {
  const jsDocParse = getParser(originalParser.parse, parserName) as any;

  const jsDocPreprocess = (text: string, options: prettier.ParserOptions) => {
    const tsPluginParser = findPluginByParser(parserName, options);

    if (!tsPluginParser) {
      return originalParser.preprocess
        ? // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          //@ts-ignore
          originalParser.preprocess(text, options)
        : text;
    }

    const preprocess = tsPluginParser.preprocess || originalParser.preprocess;

    Object.assign(parser, {
      ...parser,
      ...tsPluginParser,
      preprocess: jsDocPreprocess,
      parse: jsDocParse,
    });

    return preprocess ? preprocess(text, options) : text;
  };

  const parser = {
    ...originalParser,
    preprocess: jsDocPreprocess,
    parse: jsDocParse,
  };

  return parser;
}

export { languages, options, parsers, defaultOptions };
