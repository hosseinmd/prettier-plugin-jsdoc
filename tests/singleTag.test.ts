import * as prettier from "prettier";
import { AllOptions } from "../src/types";

function subject(code: string, options: Partial<AllOptions> = {}) {
  return prettier.format(code, {
    plugins: ["prettier-plugin-jsdoc"],
    jsdocSpaces: 1,
    parser: "babel",
    ...options,
  } as AllOptions);
}

test("single tag", async () => {
  const result = await subject(`
  /**
* @param {  string   }    param0 description
   */
function fun(param0){}

  export const SubDomain = {
    /**
     * @returns {import('axios').AxiosResponse<import('../types').SubDomain>}
     */
    async subDomain(subDomainAddress) {
    },
  };
  
`);

  expect(result).toMatchSnapshot();
});
