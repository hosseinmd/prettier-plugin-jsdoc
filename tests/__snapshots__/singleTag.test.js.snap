// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single tag 1`] = `
"/** @param {string} param0 Description */
function fun(param0) {}

export const SubDomain = {
  /** @returns {import(\\"axios\\").AxiosResponse<import(\\"../types\\").SubDomain>} */
  async subDomain(subDomainAddress) {},
};
"
`;
